### DISCUSSION TITLE
Out of memory group by?

### DISCUSSION THREAD

**Discussion by rikuiki** [duckdb/duckdb#6696] [General]
Hi, I have large table which doesn't fit ram, and contain many duplicate values.

I am trying to run something following:

select v, count(*) from t group by 1 order by 2 desc limit 10;

and get out of memory error.
I am wondering, is it because DuckDB doesn't support out of memory group by? Or it is a bug?
Is DuckDB's one of the goals to run queries on datasets which don't fit memory?

**Comment by l1t1** [comment]
please post create table and insert sql

**Comment by rikuiki** [comment]
Table has been created by importing data from external tsv files which have 3B rows.

create table t as
select * from read_csv(
    '/usr/proj/data/tsv-*',
    delim='\t', header=False,
    columns={'v': 'text', 'v2': 'text', 'v3': 'text', 's': 'bigint'});

**Comment by Mytherin** [comment]
If the hash table does not fit in memory then this query is indeed expected to fail with an OOM exception.

**Comment by rikuiki** [comment]
Thank you, I hoped duck db supports some kind of external group by when hash table doesn't fit into the memory.

**Reply by Mytherin** [reply]
It is on the roadmap but not yet supported

**Reply by rikuiki** [reply]
I am very interested in this feature. Maybe there is some feature request I can subscribe and be notified(even if it is 2 years from now)?

**Reply by Mytherin** [reply]
This should now be supported by the latest DuckDB version

**Comment by felix-hh** [comment]
Is there any workaround to grouping by a high cardinality column? I have noticed I can sort by the high cardinality column with the available memory but I'm not sure how to leverage that.

**Reply by Mytherin** [reply]
The latest DuckDB version should support out of core aggregates

**Reply by rikuiki** [reply]
Thank you for working on this.

I tested this on 1B records/100M cardinality, and it works very well, and finishes within minutes.

However, when I increased size to 10B records/1B cardinality, it looks there is some degradation:

- query didn't complete after 24h hours of waiting
- when I look at the process, it has 100% load of 7 cores out of 30, and don't produce any IO
- process consumes 23GB of ram, while I set max memory to 16GB

I am wondering if this looks like a bug?

I use following code to test it:

.open t
.timer on

SET memory_limit='16GB';
SET threads TO 30;

create table t as
select
  cast(v.range % 1000000000 as varchar) v
from range(10000000000) v;

create table t2 as select v, count(*) from t group by 1;
